### PR TITLE
make the snapshotter a self-configuring singleton

### PR DESCRIPTION
**Pull request by DominikGuzei**
This PR improves the snapshotting feature by making the following changes:

* `Space.eventSourcing.Snapshotter` is now a self-configuring singleton
* snapshots are EJSON serialized so that aggregates can be correctly de-serialized from complex snapshots. This is necessary, so that we can have entities / value objects in our aggregates :wink:

### CODE REVIEW

File: source/server/module.coffee [hunk 1]
```diff
@@ -14,6 +13,9 @@ class Space.eventSourcing extends Space.Module
       snapshotting: {
         enabled: ['SPACE_ES_SNAPSHOTTING_ENABLED', true, 'bool']
         frequency: ['SPACE_ES_SNAPSHOTTING_FREQUENCY', 10, 'int']
+      },
+      mongo: {
```
Comment (rhyslbw) [inline source/server/module.coffee]: Yes this is good to clarify now. My original intent was make it forward compatible with being able to separate commits from snapshots, but that's early optimisation

File: tests/infrastructure/replaying-projections.tests.coffee [hunk 2]
```diff
@@ -65,7 +64,7 @@ describe 'Space.eventSourcing - replaying projections', ->
       SecondCollection.remove {}
       @event = new TestEvent sourceId: 'test123', value: 'test'
       @app = new TestApp()
-      @app.configure { appId: 'TestApp' }
+      @app.reset()
```
Comment (rhyslbw) [inline tests/infrastructure/replaying-projections.tests.coffee]: Nice. Will be good to remove the appId from the Configuration as it should be set in the constructor if at all. There's an id of the running meteor process we can utilise to for abstract level identification.

File: source/server/infrastructure/snapshotter.coffee [hunk 3]
```diff
@@ -1,27 +1,52 @@
...
+  }
+
+  collection: null
   _versionFrequency: 0
 
-  constructor: (config) ->
-    @_collection = config.collection
-    @_versionFrequency = config.versionFrequency
+  onDependenciesReady: ->
+    @_setupSnapshotting() if @configuration.eventSourcing.snapshotting.enabled
```
Comment (rhyslbw) [inline source/server/infrastructure/snapshotter.coffee]: I would still leave the check against `@configuration.eventSourcing.snapshotting.enabled` in Space.eventSourcing module, as the Snapshotter shouldn't know if it's needed by the module.

File: source/server/module.coffee [hunk 4]
```diff
@@ -29,15 +31,14 @@ class Space.eventSourcing extends Space.Module
     'Space.eventSourcing.CommitPublisher'
     'Space.eventSourcing.Repository'
     'Space.eventSourcing.Projector'
+    'Space.eventSourcing.Snapshotter'
```
Comment (rhyslbw) [inline source/server/module.coffee]: This moves down to onInitialize to check if @Configuration.eventSourcing.snapshotter.enabled

File: source/server/infrastructure/snapshotter.coffee (hunk 3 above)
Comment (DominikGuzei) [inline source/server/infrastructure/snapshotter.coffee]: Yeah i also thought about that – actually we should only add the snapshotter singleton if this option is enabled, thus we would have to remove it from the `Singletons` array and conditionally map it ourselves

Comment (rhyslbw) [inline source/server/infrastructure/snapshotter.coffee]: Yep. That property would be better called `bootstrapSingletons` or something similar

**Comment by DominikGuzei** [comment]
@rhyslbw please review, there are multiple api changes now which you can also checkout in the latest donations app PR. Im still not 100% happy with the aggregate API but we are getting closer. Im just not sure if this is highest priority atm.

**Pull request merged**